### ISSUE TITLE
Cannot find imported modules (Angular)

### ISSUE THREAD

**Issue by jtsom** [atom/ide-typescript#43]
Opening up a new Angular application, created with the Angular CLI causes many Diagnostic messages saying it can't find any of the modules - `@angular/core`, `@angular/platform-browser`, etc.

<img width="943" alt="screen shot 2017-09-25 at 8 02 07 am" src="https://user-images.githubusercontent.com/13488/30807448-e0c8b468-a1c7-11e7-866a-7b788db07e64.png">
<img width="308" alt="screen shot 2017-09-25 at 8 02 17 am" src="https://user-images.githubusercontent.com/13488/30807449-e0cdc5de-a1c7-11e7-8d85-a2396e22f4bb.png">

**Comment by nickbclifford** [comment]
Looks like it can find all the other npm packages that I use in my Angular projects, so my guess would be that it's something to do with scoped npm packages. For now, I'll have to switch back to `atom-typescript` until this issue is resolved.

**Comment by yubaoquan** [comment]
I met the same problem. 

What's more, it also reports that some file is not a module but actually the file is module. 

And some  “property does not exist on value of type” errors. Consider such situation:
```
someObj = someFn()
if (someChecker.isTypeFoo(someObj)) {
    const propertyPOfFoo = someObj.p // will got an error: property p doesn't exist on value of ...
}
```
However, atom-typescript doesn't report those errors.

**Comment by wyqydsyq** [comment]
![image](https://user-images.githubusercontent.com/621277/31757088-baa9c1f6-b4ea-11e7-9a33-a96fd1929fe5.png)
I've also experienced this with the `uuid` package. I have the package itself and it's `@types/` package installed, my app runs perfectly (`tsc itself can recognize the dependency`) but `ide-typescript` incorrectly reports an error.

**Comment by wyqydsyq** [comment]
![image](https://user-images.githubusercontent.com/621277/31757803-e58d200e-b4ed-11e7-8a12-e23580828329.png)
Happens for `react` and `ramda` in another project as well with a slightly different message, even though both are very definitely installed, and both `tsc` and `atom-typescript` do not report these incorrect errors.

**Comment by wyqydsyq** [comment]
In the case of that second project I just realized that the `package.json` and `node_modules` is located above the path that the "Project Folder" in Atom is open, so it looks like perhaps `ide-typescript` isn't looking above that path e.g.
```
./
  - node_modules ← This is where my modules are installed
  - package.json
  - src/
  - apps/
    - myapp/ ← This is the active "Project Folder" in Atom
        - index.tsx ← This is where I'm editing
```

**Comment by damieng** [comment]
There is definitely a 1-to-1 mapping between the folder you open in Atom and the folder the language server takes as it's initial project folder.

Does it work if you open `./` in your example?

**Comment by wyqydsyq** [comment]
![image](https://user-images.githubusercontent.com/621277/31804970-9e2ab434-b5a0-11e7-83b8-67400434733b.png)
Opening the Node.js (where `package.json`, `tsconfig.json` and `node_modules` are) project context as the Atom Project Folder works like in the screenshot above.

`atom-typescript` is able to correctly use the right context in this scenario so I feel like the issue is that this plugin is sending the Atom Project Folder path as the context to the language server, when in reality it should be sending the Node.js project path to the language server

**Comment by damieng** [comment]
There's all sorts of issues to consider when we do this;

1. Do we stop at the first one up the tree or keep going?
2. Do we scan down the tree in every folder for additional possible tsconfig.json's?
3. Do we fire up an entire language server for every match we find both up and down?

I suspect this is going to be a per-ide-package process we have to consider and whether there is a combination that will satisfy the majority of people (and how do we handle the less popular scenarios)

**Comment by wyqydsyq** [comment]
I think what you mentioned in `1` is the behaviour that `tsc` and `atom-typescript` both appear to follow, it is also similar to how Node.js normally behaves with looking for `package.json` and `node_modules/` when resolving modules.

As far as I can tell, `atom-typescript` (which works perfectly for my use case) only ever has one "active" `tsconfig.json` at any given time, which can be opened by clicking the little `.` in the status bar: 
![image](https://user-images.githubusercontent.com/621277/31868392-9eda5c52-b7e1-11e7-87c4-194a9b122318.png)

**Comment by CodinCat** [comment]
I have the same problem in a JavaScript (not TS) project when `@ts-check` is enabled

![2017-10-27 4 31 20](https://user-images.githubusercontent.com/3193344/32094976-273b9010-bac7-11e7-8a22-7552a93fb968.png)

**Comment by tudorbarbuhc** [comment]
Nothing yet on this one?

**Comment by wyqydsyq** [comment]
`atom-typescript`, which works fine in this scenario, appears to work by only using a single `tsconfig.json`, based on the "current file" (focused tab) for context at any given time.

So I imagine `ide-typescript` would function just as well by operating in the same way.

Only send the `tsconfig.json` for the current active file to the language server. When the user changes context (e.g. opens another file, switches tabs, focuses a file in another pane) evaluate the path of the `tsconfig.json` for the new current file, if it's different to the previous one then pass the path for the new one to the language server and ideally if possible, get the language server to unload the previous one.

The only potential downside I see to this approach is that it would prevent users from seeing errors from files in a different `tsconfig.json` context to the currently active file, which I think would be preferable anyway because seeing errors not relevant to the current context would be confusing. You wouldn't see those errors from running the active file via `tsc` so I shouldn't see them in Atom by opening that file in a tab.

**Comment by nickbclifford** [comment]
Any updates on this?

**Comment by BrianDGLS** [comment]
I have this problem when opening a atom with `atom .`, however it works correctly when atom is opened with `atom ./`.

I would expect both to have the same outcome.

**Comment by wyqydsyq** [comment]
There should be a work-around for this in the next update: https://github.com/atom/ide-typescript/issues/55#issuecomment-349431488

With that update you can disable the warnings from this plugin and use the old `atom-typescript` plugin to get your TypeScript warnings, this may result more memory/cpu consumption as the `tsc` compiler might be getting ran twice, but at least you should be getting accurate tsc warnings while also getting all the nice developer experience features provided by `ide-typescript`.

**Comment by BrianDGLS** [comment]
My understanding is that in order to fix this issue `ide-typescript` (this project) needs to find the `tsconfig.json` file using the same approach as both `tsc` and `atom-typescript`.

Is this a large fix? Why can it not be included in the next release? Will existing projects suffer from this change?

I also feel that the above work-around does not address the issue. I did not install `ide-typescript` to use along with `atom-typescript` I installed it to use as a standalone Typescript ide.

**Comment by Jbz797** [comment]
Fully agree with you @BrianDGLS

**Comment by wyqydsyq** [comment]
@BrianDGLS it addresses the issue by providing a band-aid means to achieve the same end result (accurate TS warnings in your `Diagnostics` tab while still getting all the IDE features of `ide-typescript`) as resolving the issue would. Obviously it's a suboptimal solution because it requires use of an extra plugin but it does provide an effective work-around until a proper fix is implemented. It *addresses* the issue, but doesn't *resolve* it.

**Comment by nickbclifford** [comment]
Have there been any language server or `ide-typescript` updates that resolve this issue other than the (rather hacky, might I add) work-around?

**Comment by dctalbot** [comment]
Disabling and re-enabling both `ide-typescript` and `atom-typescript` solved this error for me

**Comment by tylerFowler** [comment]
Possibly fixed by [this](https://github.com/sourcegraph/javascript-typescript-langserver/issues/208)?

**Comment by BrianDGLS** [comment]
@tylerFowler That fix was merged June 3 2017. This issue is more recent than that.

**Comment by damieng** [comment]
This project is using an older version of the upstream server so that fix
night work. This needs to be bumped from 2.7.0 to 2.9.3

**Comment by daviwil** [comment]
I'll take a look at getting the TypeScript language server dependency updated this week

**Comment by DrMiaow** [comment]
+1 Same issue..

**Comment by damieng** [comment]
Possibly fixed with the new ide-typescript 0.8.0 update - let us know.

**Comment by DrMiaow** [comment]
> Possibly fixed with the new ide-typescript 0.8.0 update - let us know.

No change for me.

**Comment by BruneXX** [comment]
Hi guys, any solution to this?

> Disabling and re-enabling both `ide-typescript` and `atom-typescript` solved this error for me

but.. without that you're not able to see the linter errors in real time..right? is there no solution to this already?
